lib/shared/src/event/log_event/mod.rs

```diff
@@ -0,0 +1,828 @@
+///
+/// Transparently, as a normal [`BTreeMap`](std::collections::BTreeMap):
+///
+/// ```rust
+/// use shared::{event::*, lookup::*};
+/// let mut event = LogEvent::default();
+/// event.insert(String::from("foo"), 1);
+/// assert!(event.contains("foo"));
+/// assert_eq!(event.get("foo"), Some(&Value::from(1)));
+/// ```
+///
+/// Using remap-style lookups:
+///
+/// ```rust
+/// use shared::{event::*, lookup::*};
+/// let mut event = LogEvent::default();
+/// let lookup = LookupBuf::from_str("foo[0].(bar | bat)").unwrap();
+/// event.insert(lookup.clone(), 1);
+/// assert!(event.contains(&lookup));
+/// assert_eq!(event.get(&lookup), Some(&Value::from(1)));
+/// ```
+///
+/// It's possible to access the inner [`BTreeMap`](std::collections::BTreeMap):
+///
+/// ```rust
+/// use shared::{event::*, lookup::*};
+/// let mut event = LogEvent::default();
+/// event.insert(String::from("foo"), 1);
+///
+/// use std::collections::BTreeMap;
+/// let _inner: &BTreeMap<_, _> = event.inner();
+/// let _inner: &mut BTreeMap<_, _> = event.inner_mut();
+/// let inner: BTreeMap<_, _> = event.take();
+///
+/// let event = LogEvent::from(inner);
+/// ```
+///
+/// There exists a `log_event` macro you may also utilize to create this type:
+///
+/// ```rust
+/// use shared::{log_event, event::*, lookup::*};
+/// let event = log_event! {
+///     "foo" => 1,
+///     LookupBuf::from_str("bar.baz").unwrap() => 2,
+/// }.into_log();
+/// assert!(event.contains("foo"));
+/// assert!(event.contains(Lookup::from_str("foo").unwrap()));
+/// ```
+#[derive(PartialEq, Debug, Clone, Default, Serialize, Deserialize)]
+pub struct LogEvent {
+    #[serde(flatten)]
+    fields: BTreeMap<String, Value>,
+}
+
+impl LogEvent {
+    /// Get an immutable borrow of the given value by lookup.
+    ///
+    /// ```rust
+    /// use shared::{log_event, event::*, lookup::*};
+    /// let plain_key = "foo";
+    /// let lookup_key = LookupBuf::from_str("bar.baz").unwrap();
+    /// let event = log_event! {
+    ///     plain_key => 1,
+    ///     lookup_key.clone() => 2,
+    /// }.into_log();
+    /// assert_eq!(event.get(plain_key), Some(&Value::from(1)));
+    /// assert_eq!(event.get(&lookup_key), Some(&Value::from(2)));
+    /// ```
+    pub fn get<'a>(&self, lookup: impl Into<Lookup<'a>> + Debug) -> Option<&Value> {
+        let mut working_lookup = lookup.into();
+        let span = trace_span!("get", lookup = %working_lookup);
+        let _guard = span.enter();
+
+        // The first step should always be a field.
+        let this_segment = working_lookup.pop_front().unwrap();
+        // This is good, since the first step into a LogEvent will also be a field.
+
+        // This step largely exists so that we can make `cursor` a `Value` right off the bat.
+        // We couldn't go like `let cursor = Value::from(self.fields)` since that'd take the value.
+        match this_segment {
+            Segment::Coalesce(sub_segments) => {
```

Comment:
I was thinking about this too! At the start I was just using a vector and iterating, but it ended up being that there was need to reconstruct the lookup in some error paths like coalesce where we attempt to insert into a primitive and need to do some backtracking and fixups.

Reply:
`remap_lang::Path` has these:

https://github.com/timberio/vector/blob/3b16679f07193b05fb5f9bd69b40b4b1e4bb2dd2/lib/remap-lang/src/path.rs#L65-L135

Maybe we could offer a similar API?

Reply:
Yeah, that's similar to what I had in mind, just instead of `.foo.bar.(baz | qux)` turning into `[[".foo"], [".bar"], [".baz", ".qux"]]`, it'd let you iterate over `[[".foo", ".bar", ".baz"], [".foo", ".bar", ".qux"]]`. Both expose that there are multiple options, but the second does the work of materializing them for you. There is the theoretical danger of blowups, but real-world lookups should not be complex enough to cause problems (and being a lazy iterator would help).

I've been trying to work out if/when there's a difference in behavior between doing that kind of full materialization for paths with coalesces vs going segment-by-segment with backtracking. It _seems_ like there could be in the insert case at least, but it's complicated enough that I'm not sure.

Reply:
Yeah, maybe it's better to leave this for future perf exploration?